### ISSUE TITLE
Overruling rules (AS announcements)

### ISSUE THREAD

**Issue by marcgweg** [FORTH-ICS-INSPIRE/artemis#288]
**Describe the bug**
I have a list of prefixes with the same origin AS. I want to create two rulesets:
1) prefixlist with origin AS and a select set of neighbors;
2) prefixlist with origin AS and an other select list of neighbors for which a no_export is required.
Aspected behaviour is that both use-cases is respected.

**Affected Component(s)**
- [X] Back-End (Database, Microservices, Containers, etc)
- [ ] Front-End (Flask, API, etc)
- [ ] Docs
- [ ] Build System

**To Reproduce**
In config.yaml I have the following setup:
prefixes:
  simple_prefix_list: &my_prefixes
  - 192.168.78.0/24
  - 192.168.10.0/24
<... skip monitors ...>
asns:
  my_asn: &my_asn
  1
  my_neighbors: &myneighbors
  - 11
  - 12
  my_neighbors: &my_noexp_neighbors
  - 21
  - 222
rules:
- prefixes:
  - *my_prefixes
  origin_asns:
  - *my_asn
  neighbors:
  - *my_neighbors
  mitigation: manual
- prefixes:
  - *my_prefixes
  origin_asns:
  - *my_asn
  neighbors:
  - *my_noexp_neighbors
  policies:
  - 'no-export'
  mitigation: manual

**Expected behavior**
Both situation will be observed in their respective situation (is the combination origin AS and neighbour with given policy).
Now only the no-export rule is followed.

**Screenshots**

**System (please complete the following information):**
 - Version latest

**Additional context**

**Comment by vkotronis** [comment]
@marcgweg thank you for reporting this Marc. Could you please elaborate more on the issue, so that we can add respective tests and see what is going on? I.e., do you actually see BGP updates within `*my_prefixes` with origin `*my_asn` and neighbor within `myneighbors` and only the second rule kicks in? Could you also please fix the typo in `my_neighbors`? It seems that you use exactly the same name for 2 different sets (the normal and the no export ones) and the yaml anchors seem also mistyped: `&myneighbors` (instead of `&my_neighbors`). This does not generate an error in yaml but may cause this logical problem with the 2 rules. After this is fixed, could you try again and let us know? if it is still an issue I will add more tests to check what is going on. Thanks!

**Comment by vkotronis** [comment]
To help clarify my previous comment, please try the following conf:
```
simple_prefix_list: &my_prefixes
    - 192.168.78.0/24
    - 192.168.10.0/24
    <... skip monitors ...>
    asns:
    my_asn: &my_asn
    1
    my_neighbors: &my_neighbors
    - 11
    - 12
    my_noexp_neighbors: &my_noexp_neighbors
    - 21
    - 222
    rules:
    prefixes:
        *my_prefixes
        origin_asns:
        *my_asn
        neighbors:
        *my_neighbors
        mitigation: manual
    prefixes:
        *my_prefixes
        origin_asns:
        *my_asn
        neighbors:
        *my_noexp_neighbors
        policies:
        'no-export'
        mitigation: manual
```
(with the mentioned typos fixed). If the problem persists, please let me know.

**Comment by marcgweg** [comment]
Okay, I use a different config in production, then I sent to you. In your config I miss the "-" symbols. I check my config, and try to get results you asked for.

**Comment by marcgweg** [comment]
I guess the following was wrongly typed:
my_noexp_neighbors: &my_noexp_neighbors should be
my_neighbors: &my_noexp_neighbors !?

**Comment by slowr** [comment]
@marcgweg the names you specify the AS groups and prefixes (my_asn, my_neighbors, etc) are like variables so they need to be unique. For your example this is the correct configuration:
```
prefixes:
    simple_prefix_list: &simple_prefix_list
        - 192.168.78.0/24
        - 192.168.10.0/24

<... skip monitors ...>

asns:
    my_asn: &my_asn
        8522
    my_neighbors_1: &my_neighbors_1
        - 11
        - 12
    my_neighbors_2: &my_neighbors_2
        - 21
        - 22

rules:
- prefixes:
    - *simple_prefix_list
  origin_asns:
    - *my_asn
  neighbors:
    - *my_neighbors_1
  mitigation:
      manual
- prefixes:
    - *simple_prefix_list
  origin_asns:
    - *my_asn
  neighbors:
    - *my_neighbors_2
  policies:
     'no-export'
  mitigation:
     manual
```

**Comment by marcgweg** [comment]
AH, okay. I thought the my_neighbors(_1) in the construction:
my_neighbors_1: &my_neighbors_1
        - 11
        - 12
was kind of a type declaration. So every neighbor should be unique in both left and right hand side...

**Comment by marcgweg** [comment]
okay, I have now the following config:

prefixes:
    simple_prefix_list_1: &prefix_list_1
        - 192.168.78.0/24
        - 192.168.10.0/24

# skip monitors

asns:
    my_asn_1: &my_asn_1
        8522
    my_neighbors_1: &my_neighbors_1
        - 11
        - 12
    my_neighbors_2: &my_neighbors_2
        - 21
        - 22

rules:
- prefixes:
    - *prefix_list_1
  origin_asns:
    - *my_asn_1
  neighbors:
    - *my_neighbors_1
  mitigation:
      manual
- prefixes:
    - *prefix_list_1
  origin_asns:
    - *my_asn
  neighbors:
    - *my_neighbors_2
  policies:
     'no-export'
  mitigation:
     manual

**Comment by marcgweg** [comment]
Btw: I have 18 ASNs in the my_neighbors_2 list (with the no-export option)

**Comment by vkotronis** [comment]
No problem about the edit modus. So now everything works as expected, right?

**Comment by marcgweg** [comment]
Oh sorry, was blaming the Git edit modus.
But when I try the edit the config within Artemis, I get the strange behaviour that after pressing save the first time, in 3 iterations the config gets scrambled up. When I do the edit again, scratch everything, and paste the config again --> then press save, the strange behaviour doesn't return. Another issue I guess...

**Comment by marcgweg** [comment]
I still get the issue when the origin AS8522 with neighbor 11 gets a hijack, when 2 more Ases are seen within the path. As like it is seen as a no-export path.

**Comment by vkotronis** [comment]
I do not get what you mean with scrambling, do you have an example? When you press save, we format the conf in fully-compliant (i.e., format-wise) yaml that remains stable when diffs are created (e.g., during a "learn hijack rule" event). If possible please create a new issue with the exact erroneous behavior so that we reproduce it (and make sure that the formatter works always correctly).

So what is going on in your case, is that the first rule you have for the prefix clears the BGP update (since the origin and neighbor is valid), but the second one activates route leakage check. This is as intended in the design of the detection mechanism (https://github.com/FORTH-ICS-INSPIRE/artemis/blob/master/backend/core/detection.py#L811 ). So what you actually need is that the first rule clears the BGP update and the second one is not activated at all, correct? If yes, this can be done yes, but need to check if it affects sth else in the design (just want  to understand the problem correctly).

**Comment by vkotronis** [comment]
So what we need to do there is probably simply return when the first rule clears the hijack, but maybe other use cases require all rules related to the prefix be checked first. When we designed this, we were conservative to avoid missing a hijack, but I think I get your use case now. Need to coordinate with the other 2 devs in case I am missing sth.

**Comment by marcgweg** [comment]
ok. Thank. My use case is a scenario of anycast addresses, which are deployed globally (hence the first neighbors block) and locally (the no-export group). The global group, I wish to get notified in case of a hijack. The second group (no-export) in case of unwanted route leakage.

**Comment by marcgweg** [comment]
Th scrabble case: I cannot reproduce now... Perhaps the docker-compose pull was the solution to this.

**Comment by marcgweg** [comment]
As a temporary workaround, I have removed the no-export rule group from the config. Guess it is now becoming a feature request instead of a bug report!?

**Comment by vkotronis** [comment]
No it is fine as a bug report; I think the root cause is that since we examine the different dimensions of a hijack event in parallel, it may happen that we do not synchronize the analysis for exactly the same rules in case there are multiple rule for exactly the same prefix, like in your use case. So I think it is a bug in that sense; we will fix it for the coming release (will also notify when it is available in the bleeding edge master). I will take a careful look at it next week to root it out, and will also add some tests related to this use case to make sure that we deal with the bug. Many thanks for reporting this Marc, it seems to be insidious, not covered by the rest of the tests we put in place :) .

**Comment by vkotronis** [comment]
Note to my self for dealing with this issue (for dev purposes): passing the indexed conf object/rule instead of the initial prefix node (https://github.com/FORTH-ICS-INSPIRE/artemis/blob/master/backend/core/detection.py#L460) in the respective parallel checkers, so that they all work on exactly the same object at exactly the same time and avoid rule confusion.

**Comment by vkotronis** [comment]
@marcgweg the fix will be included in latest master ("latest" tag) in a few hours. It will also be part of the 1.4.1 release, but if you would like to upgrade to latest (pulling the latest code via `git`, and containers via `docker-compose pull`) we would welcome testing and making sure that it works exactly as we expect (just make sure you also pull the latest containers themselves)

**Comment by marcgweg** [comment]
@vkotronis Just installed the latest version. I don't get the previous error situation. So the check seems correct. I'll try to contact one of the no-export ASNs to temporary leak the route information. Thanks for the handling of this issue.

**Comment by vkotronis** [comment]
I am simply re-opening the issue to first get the final OK by Marc (controlled route leak check).

**Comment by marcgweg** [comment]
I tested the functionality by placing a correct neighbor in the no-export group. I got directly a hijack message of type L (indicating rout leakage). So it works as suspected.
